tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/GraphFunction.java

```diff
@@ -0,0 +1,301 @@
+import org.tensorflow.op.Ops;
+import org.tensorflow.proto.framework.AttrValue;
+import org.tensorflow.proto.framework.TensorInfo;
+
+/**
+ * A Function created from a TensorFlow {@link Graph}. Callable in Graph or Eager mode via the {@link
+ * Ops#callFunction(GraphFunction, Operand[])} ops or the {@link #call(Ops, Operand[])} methods.
+ *
+ * <p>By default, resolves arguments and returns outputs by position.
+ * Names can be specified for the inputs and outputs (and used when calling) using {@link NamedGraphFunction} and {@link
+ * #create(Graph, String, boolean, Map, Map, String)}.
+ *
+ * Functions can also be created from {@link ConcreteFunction}s using {@link #create(ConcreteFunction, boolean)}.
+ */
+public class GraphFunction implements AutoCloseable {
+
+  private final TF_Function nativeHandle;
+  private final PointerScope scope;
+  private final int numInputs;
+  private final int numOutputs;
+
+  GraphFunction(TF_Function nativeHandle, int numInputs, int numOutputs) {
+    this.numInputs = numInputs;
+    this.numOutputs = numOutputs;
+    scope = new PointerScope();
+    this.nativeHandle = nativeHandle;
+    scope.attach(nativeHandle);
+  }
+
+  /**
+   * This causes native errors when I use it (Linux GPU, 6.1 CC), but I'm leaving it because how to enable XLA JIT is
+   * extremely non-obvious.
+   *
+   * Causes {@code OP_REQUIRES failed at xla_ops.cc:363 : Not found: could not find registered platform with id:
+   * 0x7f75af03e6e8} (it's a warning, but the resulting TF_Status fails).
+   */
+  private void makeJit() {
+    try (PointerScope scope = new PointerScope()) {
+      byte[] bytes = AttrValue.newBuilder().setB(true).build().toByteArray();
+      BytePointer trueValue = new BytePointer(bytes);
+
+      TF_Status status1 = TF_Status.newStatus();
+      TF_FunctionSetAttrValueProto(nativeHandle(), "_XlaMustCompile", trueValue, bytes.length, status1);
+      status1.throwExceptionIfNotOK();
+
+      TF_Status status2 = TF_Status.newStatus();
+      TF_FunctionSetAttrValueProto(nativeHandle(), "_noinline", trueValue, bytes.length, status2);
+      status2.throwExceptionIfNotOK();
+    }
+  }
+
+  private static TF_Operation outputHandle(Operand<?> operand) {
+    if (operand == null) {
+      throw new NullPointerException("Can't get output handle for null operand");
+    }
+
+    Pointer handle = operand.asOutput().getUnsafeNativeHandle();
+    if (handle.isNull()) {
+      throw new NullPointerException("Native handle of operand is null, has it been closed?");
+    }
+
+    if (!(handle instanceof TF_Operation)) {
+      throw new IllegalArgumentException("Operand was not a graph operand");
+    }
+
+    return (TF_Operation) handle;
+  }
+
+  private static TF_Output resolveToOutput(List<Operand<?>> operands) {
+    TF_Output handles = new TF_Output(operands.size());
+    for (int i = 0; i < operands.size(); i++) {
+      Operand<?> input = operands.get(i);
+      //TODO body.checkInput(input)
```

Comment:
can you explain more in detail here what is left TODO?

Reply:
It's from #207, I'd originally written this in that branch.

Reply:
I'm expecting it to get merged before this branch anyways, in which case I'll uncomment it.